src/Nethermind/Nethermind.JsonRpc/Modules/Eth/FeeHistory/FeeHistoryOracle.cs

```diff
@@ -1,121 +1,225 @@
+                return ResultWrapper<FeeHistoryResults>.Fail("newestBlock: Block is not available",
+                    ErrorCodes.ResourceUnavailable);
             }
 
-            long oldestBlockNumber = block!.Number;
-            Stack<UInt256> baseFeePerGas = new((int)(blockCount + 1));
-            baseFeePerGas.Push(BaseFeeCalculator.Calculate(block!.Header, _specProvider.GetSpecFor1559(block!.Number + 1)));
-            Stack<UInt256> baseFeePerBlobGas = new((int)(blockCount + 1));
-            BlobGasCalculator.TryCalculateBlobGasPricePerUnit(block!.Header, out UInt256 blobGas);
-            baseFeePerBlobGas.Push(blobGas == UInt256.MaxValue ? 0 : blobGas);
+            BlockFeeHistorySearchInfo info = historyInfo.Value;
 
-            Stack<double> gasUsedRatio = new((int)blockCount);
-            Stack<double> blobGasUsedRatio = new((int)blockCount);
+            // Assumes if blockCount is ever greater than the BlockNumber then BlockNumber must fall within integer size limits
+            var effectiveBlockCount = (info.BlockNumber < blockCount - 1) ? (int)info.BlockNumber + 1 : blockCount;
+            var tempBlockCount = effectiveBlockCount + 1;
+            ArrayPoolList<UInt256> baseFeePerGas = new(tempBlockCount, tempBlockCount);
+            ArrayPoolList<UInt256> baseFeePerBlobGas = new(tempBlockCount, tempBlockCount);
+            ArrayPoolList<double> gasUsedRatio = new(effectiveBlockCount, effectiveBlockCount);
+            ArrayPoolList<double> blobGasUsedRatio = new(effectiveBlockCount, effectiveBlockCount);
+            ArrayPoolList<ArrayPoolList<UInt256>>? rewards = rewardPercentiles is null || rewardPercentiles.Length == 0
+                ? null
+                : new ArrayPoolList<ArrayPoolList<UInt256>>(effectiveBlockCount, effectiveBlockCount);
 
-            Stack<UInt256[]>? rewards = rewardPercentiles is null || rewardPercentiles.Length == 0 ? null : new Stack<UInt256[]>((int)blockCount);
+            long oldestBlockNumber = info.BlockNumber;
+            baseFeePerGas[effectiveBlockCount] = info.BaseFeePerGasEst;
+            baseFeePerBlobGas[effectiveBlockCount] = (info.BaseFeePerBlobGas);
 
-            while (block is not null && blockCount > 0)
+            while (historyInfo is not null && effectiveBlockCount-- > 0)
             {
-                oldestBlockNumber = block.Number;
-                baseFeePerGas.Push(block.BaseFeePerGas);
-                BlobGasCalculator.TryCalculateBlobGasPricePerUnit(block!.Header, out blobGas);
-                baseFeePerBlobGas.Push(blobGas == UInt256.MaxValue ? 0 : blobGas);
-                gasUsedRatio.Push(block.GasUsed / (double)block.GasLimit);
-                blobGasUsedRatio.Push((block.BlobGasUsed ?? 0) / (double)Eip4844Constants.MaxBlobGasPerBlock);
+                info = historyInfo.Value;
+                oldestBlockNumber = info.BlockNumber;
+                baseFeePerGas[effectiveBlockCount] = info.BlockBaseFeePerGas;
```

Comment:
Not `tempBlockCount-1`?

Reply:
tempBlockCount = effectiveBlockCount + 1,
its a variable used to avoid repetition of the same calculation, just before the loop.

So here its the same thing, since thats the last element in the ArrayPoolList

Reply:
I mean you use effectiveBlockCount for 4 arrays but two of them are +1 longer seems like, just wanted to confirm that usage of effectiveBlockCount as index for all of them is ok

Reply:
LGTM

Reply:
yea it is ok.

the other two return some extra info related to the requested block.